src/main/scala/is/hail/distributedmatrix/BlockMatrixIsDistributedMatrix.scala

```diff
@@ -0,0 +1,179 @@
+  def scalarDivide(m: M, i: Double): M = map(_ / i)(m)
+  def scalarAdd(i: Double, m: M): M = map(i + _)(m)
+  def scalarSubtract(i: Double, m: M): M = map(i - _)(m)
+  def scalarMultiply(i: Double, m: M): M = map(i * _)(m)
+  def scalarDivide(i: Double, m: M): M = map(i / _)(m)
+
+  private def mapWithRowIndex(op: (Double, Int) => Double)(x: M): M = {
+    val nRows: Long = x.numRows
+    val nCols: Long = x.numCols
+    val rowsPerBlock: Int = x.rowsPerBlock
+    val colsPerBlock: Int = x.colsPerBlock
+    val rowBlocks: Int = ((nRows - 1) / rowsPerBlock).toInt + 1
+    val colBlocks: Int = ((nCols - 1) / colsPerBlock).toInt + 1
+    val rowsRemainder: Int = (nRows % rowsPerBlock).toInt
+    val colsRemainder: Int = (nCols % colsPerBlock).toInt
+    val blocks: RDD[((Int, Int), Matrix)] = x.blocks.mapValuesWithKey { case ((blockRow, blockCol), m) =>
+      new DenseMatrix(m.numRows, m.numCols, m.toArray.zipWithIndex.map { case (e, j) =>
+        val rowsInThisBlock: Int = (if (blockRow + 1 == rowBlocks && rowsRemainder != 0) rowsRemainder else rowsPerBlock)
+        val colsInThisBlock: Int = (if (blockCol + 1 == colBlocks && colsRemainder != 0) colsRemainder else colsPerBlock)
+        if (blockRow.toLong * rowsInThisBlock + j % rowsInThisBlock < nRows &&
+          blockCol.toLong * colsInThisBlock + j / rowsInThisBlock < nCols)
+          op(e, blockRow * rowsInThisBlock + j % rowsInThisBlock)
+        else
+          e
+      })
+    }
+    new BlockMatrix(blocks, x.rowsPerBlock, x.colsPerBlock, x.numRows(), x.numCols())
+  }
+  private def mapWithColIndex(op: (Double, Int) => Double)(x: M): M = {
+    val nRows = x.numRows
+    val nCols = x.numCols
+    val rowsPerBlock = x.rowsPerBlock
+    val colsPerBlock = x.colsPerBlock
+    val rowBlocks = ((nRows - 1) / rowsPerBlock).toInt + 1
+    val colBlocks = ((nCols - 1) / colsPerBlock).toInt + 1
+    val rowsRemainder = (nRows % rowsPerBlock).toInt
+    val colsRemainder = (nCols % colsPerBlock).toInt
+    val blocks: RDD[((Int, Int), Matrix)] = x.blocks.mapValuesWithKey { case ((blockRow, blockCol), m) =>
+      new DenseMatrix(m.numRows, m.numCols, m.toArray.zipWithIndex.map { case (e, j) =>
+        val rowsInThisBlock: Int = (if (blockRow + 1 == rowBlocks && rowsRemainder != 0) rowsRemainder else rowsPerBlock)
+        val colsInThisBlock: Int = (if (blockCol + 1 == colBlocks && colsRemainder != 0) colsRemainder else colsPerBlock)
+        if (blockRow * rowsInThisBlock + j % rowsInThisBlock < nRows &&
+          blockCol * colsInThisBlock + j / rowsInThisBlock < nCols)
+          op(e, blockCol * colsInThisBlock + j / rowsInThisBlock)
+        else
+          e
+      })
+    }
+    new BlockMatrix(blocks, x.rowsPerBlock, x.colsPerBlock, x.numRows(), x.numCols())
+  }
+
+  def vectorAddToEveryColumn(v: Array[Double])(m: M): M = {
+    require(v.length == m.numRows())
+    val vbc = m.blocks.sparkContext.broadcast(v)
+    mapWithRowIndex((x,i) => x + vbc.value(i))(m)
```

Comment:
Same performance comment here.  This should be vectorized (e.g. calling a vector add looping over rows).